xml/encoder_test.go

```diff
@@ -0,0 +1,447 @@
+	v := encoder.Value
+	root := v.RootElement()
+
+	// nested `payload` shape
+	o := root.Key("payload")
+
+	// nested `mapStr` shape
+	ns, closeFn0 := o.NestedElement()
+	o1 := ns.Key("mapstr")
+
+	m := o1.FlattenedMap()
+
+	e, closeFn2 := m.Entry()
+	e.Key("key").String("abc")
+	e.Key("value").Integer(123)
+	closeFn2()
+
+	closeFn0()
+
+	ex := []byte(`<payload><mapstr><key>abc</key><value>123</value></mapstr></payload>`)
+	verify(t, encoder, ex)
+}
+
+func TestEncodeMapNamed(t *testing.T) {
+	encoder := xml.NewEncoder()
+	v := encoder.Value
+	root := v.RootElement()
+
+	// nested `payload` shape
+	o := root.Key("payload")
+
+	// nested `mapStr` shape
+	ns, closeFn0 := o.NestedElement()
+	o1 := ns.Key("mapNamed")
+
+	m, closeFn1 := o1.Map()
+
+	e, closeFn2 := m.Entry()
+	e.Key("namedKey").String("abc")
+	e.Key("namedValue").Integer(123)
+	closeFn2()
+
+	closeFn1()
+	closeFn0()
+
+	ex := []byte(`<payload><mapNamed><entry><namedKey>abc</namedKey><namedValue>123</namedValue></entry></mapNamed></payload>`)
+	verify(t, encoder, ex)
+}
+
+func TestEncodeMapShape(t *testing.T) {
+	encoder := xml.NewEncoder()
+	v := encoder.Value
+	root := v.RootElement()
+
+	// nested `payload` shape
+	o := root.Key("payload")
+
+	// nested `mapStr` shape
+	ns, closeFn0 := o.NestedElement()
+	o1 := ns.Key("mapShape")
+
+	m, closeFn1 := o1.Map()
+
+	e, closeFn2 := m.Entry()
+	e.Key("key").String("abc")
+
+	//  delegate to shape encoding function
+	o2, closeFn3 := e.Key("value").NestedElement()
+	o2.Key("shapeVal").Integer(1)
+
+	// close all
+	closeFn3()
+	closeFn2()
+	closeFn1()
+	closeFn0()
+
+	ex := []byte(`<payload><mapShape><entry><key>abc</key><value><shapeVal>1</shapeVal></value></entry></mapShape></payload>`)
+	verify(t, encoder, ex)
+}
+
+func TestEncodeMapFlattenShape(t *testing.T) {
+	encoder := xml.NewEncoder()
+	v := encoder.Value
+	root := v.RootElement()
+
+	// nested `payload` shape
+	o := root.Key("payload")
+
+	// nested `mapStr` shape
+	ns, closeFn0 := o.NestedElement()
+	o1 := ns.Key("mapShape")
+
+	m := o1.FlattenedMap()
+
+	e, closeFn2 := m.Entry()
+	e.Key("key").String("abc")
+
+	//  delegate to shape encoding function
+	o2, closeFn3 := e.Key("value").NestedElement()
+	o2.Key("shapeVal").Integer(1)
+
+	// close all
+	closeFn3()
+	closeFn2()
+	closeFn0()
+
+	ex := []byte(`<payload><mapShape><key>abc</key><value><shapeVal>1</shapeVal></value></mapShape></payload>`)
+	verify(t, encoder, ex)
+}
+
+func TestEncodeMapNamedShape(t *testing.T) {
+	encoder := xml.NewEncoder()
+	v := encoder.Value
+	root := v.RootElement()
+
+	// nested `payload` shape
+	o := root.Key("payload")
+
+	// nested `mapStr` shape
+	ns, closeFn0 := o.NestedElement()
+	o1 := ns.Key("mapNamedShape")
+
+	m, closeFn1 := o1.Map()
+
+	e, closeFn2 := m.Entry()
+	e.Key("namedKey").String("abc")
+
+	//  delegate to shape encoding function
+	o2, closeFn3 := e.Key("namedValue").NestedElement()
+	o2.Key("shapeVal").Integer(1)
+
+	// close all
+	closeFn3()
+	closeFn2()
+	closeFn1()
+	closeFn0()
```

Comment:
For testing sanity maybe it would be better to put these node creations and their close in a closure so you can use defer's on the close functions?  This probably would help track down some hard to track bugs.

https://play.golang.org/p/WVFbRgLfMBF

This would make it a little bit more easier to do multiple level nested tests.